# 0.Libraray Import & Data Load
- 유동인구 데이터 (SKT)
- 카드데이터 (신한카드)
- 유통데이터 (편의점, GS)

- 외부데이터
  - 코로나데이터 (코로나 확진자 수 데이터)
  - 휴일 여부 데이터 (2020, 2019)

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
# 데이터 분석 library
import pandas as pd
import numpy as np

# 통계용 library
import scipy
from scipy.stats.stats import pearsonr # 상관계수
import statsmodels.formula.api as smf # 회귀모델용

!pip install pingouin
import pingouin as pg # 편상관분석용

# 기타
from functools import reduce
import datetime
from datetime import datetime, timedelta
from itertools import combinations
import warnings
warnings.filterwarnings('ignore')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


     |████████████████████████████████| 225kB 2.8MB/s 
  Created wheel for pingouin: filename=pingouin-0.3.8-cp36-none-any.whl size=221688 sha256=02d899219a698e08bf1bc5ec61bd40da14e9f8d5382fadc98f1034b6f6b01496
  Stored in directory: /root/.cache/pip/wheels/d6/9e/53/f885f73f29cf7c8cac3d8f4b1532bbfef2f5eb543946ac9055
  Created wheel for outdated: filename=outdated-0.2.0-cp36-none-any.whl size=4961 sha256=40149139f6d6e99e0c2cdb4f5b0a96a166b8d387874a1cde339931a61c767265
  Stored in directory: /root/.cache/pip/wheels/fd/7c/ef/814f514d31197310872b5abf353feb8fef9d67ee658e1e7e39
  Created wheel for littleutils: filename=littleutils-0.2.2-cp36-none-any.whl size=7051 sha256=efb46e551d31ef893f99e9cd986d8f7ef776525751ce02ea2bde4963f1d55e80
  Stored in directory: /root/.cache/pip/wheels/53/16/9f/ac67d15c40243754fd73f620e1b9b6dedc20492ecc19a2bae1
Successfully built pingouin outdated littleutils


## 유동인구 데이터 (2020, 2019)
- SKT 유동인구 데이터 2020, 2019 사용

In [5]:
# 2020년 데이터
skt_age_202002 = pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_202002.CSV", sep="|")
skt_age_202003 = pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_202003.CSV", sep="|")
skt_age_202004 = pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_202004.CSV", sep="|")
skt_age_202005 = pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_202005.CSV", sep="|")

# 2019 유동인구 데이터 age
skt_age_201902=pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_201902.CSV", sep="|")
skt_age_201903=pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_201903.CSV", sep="|")
skt_age_201904=pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_201904.CSV", sep="|")
skt_age_201905=pd.read_csv("/content/drive/My Drive/big_contest/skt/4개지역_FLOW_AGE_201905.CSV", sep="|")

# skt 데이터 합침 (2020, 2019)
skt_2020 = pd.concat([skt_age_202002,skt_age_202003,skt_age_202004,skt_age_202005]) # 2020년 데이터
skt_2019 = pd.concat([skt_age_201902, skt_age_201903, skt_age_201904, skt_age_201905]) # 2019년 데이터

print(skt_2020.shape, skt_2019.shape)
skt_2020

(8349, 34) (8280, 34)


STD_YM   STD_YMD  ...  WMAN_FLOW_POP_CNT_6569 WMAN_FLOW_POP_CNT_70U
0     202002  20200201  ...                 2440.61               2665.43
1     202002  20200202  ...                 1704.22               2026.77
2     202002  20200203  ...                 3070.67               3349.99
3     202002  20200204  ...                 3091.30               3494.91
4     202002  20200205  ...                 2960.73               3204.18
...      ...       ...  ...                     ...                   ...
2134  202005  20200527  ...                 1086.34               1143.37
2135  202005  20200528  ...                 1291.19               1515.36
2136  202005  20200529  ...                 1356.15               1455.89
2137  202005  20200530  ...                 1242.56               1372.38
2138  202005  20200531  ...                 1034.47               1142.95

[8349 rows x 34 columns]

In [6]:
# skt 유동인구 데이터

# 2020 유동인구 데이터 연령 총합 변수 만들기
skt_2020['sum_20'] = skt_2020.iloc[:,4:].sum(axis=1)
skt_2020 = skt_2020[['STD_YMD','HDONG_NM','sum_20']]
skt_2020.columns = ['STD_DD','HDONG_NM','sum_20']
skt_2020 = skt_2020.reset_index(drop=True)

# 2019 유동인구 데이터 연령 총합 변수 만들기
skt_2019["sum_19"] = skt_2019.iloc[:,4:].sum(axis=1)
skt_2019 = skt_2019[["STD_YMD","HDONG_NM","sum_19"]]
skt_2019.columns = ['STD_DD','HDONG_NM','sum_19']
skt_2019 = skt_2019.reset_index(drop=True)

print(skt_2020.shape, skt_2019.shape)
skt_2019

(8349, 3) (8280, 3)


STD_DD HDONG_NM     sum_19
0     20190201      소공동  242230.68
1     20190202      소공동  133411.99
2     20190203      소공동   90124.93
3     20190204      소공동   83378.55
4     20190205      소공동  111009.47
...        ...      ...        ...
8275  20190527     고산3동   47011.48
8276  20190528     고산3동   49344.74
8277  20190529     고산3동   50345.41
8278  20190530     고산3동   50148.22
8279  20190531     고산3동   52772.46

[8280 rows x 3 columns]

## 카드데이터 (2020, 2019)
- 신한카드 데이터 (외국인 + 내국인)
- 신한카드 행정동 데이터
- 카드 데이터 + 행정동 데이터 Merge
- 2020, 2019 데이터 분리
- 카드데이터 업종별 대분류 리스트 변수 생성

In [7]:
# 신한 카드 데이터 (내국인, 외국인)
card_re = pd.read_csv('/content/drive/My Drive/big_contest/shinhan/CARD_SPENDING_RESIDENT.txt', sep = "\t", engine='python', encoding = "cp949")
card_fo = pd.read_csv('/content/drive/My Drive/big_contest/shinhan/CARD_SPENDING_FOREIGNER.txt', sep = "\t", engine='python', encoding = "cp949")

# 신한 카드 행정동 데이터
dong = pd.read_excel('/content/drive/My Drive/big_contest/shinhan/02_혁신아이디어분야_카드소비데이터(신한카드)_데이터정의서.xlsx', sheet_name='(참고) 행정동코드',index_col=0, header=None)

print(card_re.shape, card_fo.shape, dong.shape)

(2537753, 8) (238829, 7) (71, 4)


In [8]:
# 카드 행정동 데이터 전처리
dong.columns = ['GU_CD','DONG_CD','GU_NM','HDONG_NM']
dong.reset_index(inplace=True, drop=True)
dong = dong.loc[2:]
dong = dong.reset_index(drop=True)

# 컬럼 type int 형으로 변경
dong['GU_CD'] = dong['GU_CD'].astype(int)
dong['DONG_CD'] = dong['DONG_CD'].astype(int)

# 카드 데이터 (내국인 + 외국인) concat
card = pd.concat([card_fo,card_re])

# 행정동 이름과 카드 데이터 merge
card = pd.merge(card, dong, on=['GU_CD','DONG_CD'])

print(card.shape)
card

(2776582, 11)


STD_DD  GU_CD  DONG_CD  MCT_CAT_CD  ... SEX_CD  AGE_CD   GU_NM HDONG_NM
0        20190204    260      510          81  ...    NaN     NaN  대구 수성구     범어1동
1        20190213    260      510          40  ...    NaN     NaN  대구 수성구     범어1동
2        20190416    260      510          40  ...    NaN     NaN  대구 수성구     범어1동
3        20190419    260      510          42  ...    NaN     NaN  대구 수성구     범어1동
4        20190308    260      510          10  ...    NaN     NaN  대구 수성구     범어1동
...           ...    ...      ...         ...  ...    ...     ...     ...      ...
2776577  20200531    350      710          81  ...      M    45.0  서울 노원구     상계9동
2776578  20200531    350      710          81  ...      M    50.0  서울 노원구     상계9동
2776579  20200531    350      710          81  ...      M    55.0  서울 노원구     상계9동
2776580  20200531    350      710          81  ...      M    60.0  서울 노원구     상계9동
2776581  20200531    350      710          81  ...      M    65.0  서울 노원구     상계9동

[2776582 rows x 11 columns]

In [9]:
# 일별, 구별 ,동별 USE_AMT, USE_AMT 합
card_group = card.groupby(["STD_DD",'GU_NM','HDONG_NM'], as_index=False)['USE_CNT','USE_AMT'].agg("sum")

# 2020, 2019 카드 데이터 분리
card_2020 = card_group[card_group.STD_DD>=20200101] # 2020년 카드 데이터
card_2019 = card_group[card_group.STD_DD<20200101] # 2019년 카드 데이터

print(card_2020.shape, card_2019.shape)

(8349, 5) (8280, 5)


In [10]:
# 카드 대분류 list dictionary
type_list = {'10':'숙박(10)', '20':'레저용품(20)', '21':'레저업소(21)', 
            '22':'문화취미(22)', '30':'가구(30)' , '31':'전기(31)' ,
            '32':'주방용구(32)', '33':'연료판매(33)' , '34':'광학제품(34)',
            '35':'가전(35)', '40':'유통업(40)', '42':'의복(42)', 
            '43':'직물(43)', '44':'신변잡화(44)', '50':'서적문구(50)',
            '52':'사무통신(52)', '60':'자동차판매(60)','62':'자동차정비(62)',
            '70':'의료기관(70)', '71':'보건위생(71)', '80':'요식업소(80)',
            '81':'음료식품(81)', '92':'수리서비스(92)'}

# 업종 list
MCT_CAT_CD_list = card['MCT_CAT_CD'].unique()
MCT_CAT_CD_list = sorted(MCT_CAT_CD_list)
MCT_CAT_CD_list = list(map(str, MCT_CAT_CD_list))
print(len(MCT_CAT_CD_list), MCT_CAT_CD_list)

23 ['10', '20', '21', '22', '30', '31', '32', '33', '34', '35', '40', '42', '43', '44', '50', '52', '60', '62', '70', '71', '80', '81', '92']


## 편의점(유통, GS) 데이터 (2020,2019)
- GS 제공 편의점 데이터
- 구조방정식, 매개효과용 데이터 변수 : gs
- 회귀용 데이터 변수명 : gs_row -> 2020, 2019 분리해서 사용

In [11]:
# GS 데이터 가져오기
gs = pd.read_excel('/content/drive/My Drive/big_contest/GS/유통(GS리테일).xlsx', sheet_name='종합테이블(TB_DD_STR_PURCH_CLS)')
print(gs.shape)
gs

(14460, 13)


OPER_DT    PVN  BOR  ADMD  ...  LCLS_50_P  LCLS_60_P  LCLS_70_P  LCLS_80_P
0      20190201  대구광역시  수성구  고산1동  ...   0.044679   0.005075   0.004820   0.003637
1      20190201  대구광역시  수성구  고산2동  ...   0.034056   0.000844   0.008893   0.000913
2      20190201  대구광역시  수성구  고산3동  ...   0.029606   0.009117   0.008831   0.005697
3      20190201  대구광역시  수성구   두산동  ...   0.134191   0.000000   0.021196   0.000000
4      20190201  대구광역시  수성구  만촌1동  ...   0.035427   0.005284   0.005734   0.002719
...         ...    ...  ...   ...  ...        ...        ...        ...        ...
14455  20200531  서울특별시   중구   중림동  ...   0.070094   0.006347   0.013541   0.004203
14456  20200531  서울특별시   중구   청구동  ...   0.076963   0.001592   0.000353   0.005955
14457  20200531  서울특별시   중구    필동  ...   0.079408   0.007826   0.008298   0.004236
14458  20200531  서울특별시   중구   황학동  ...   0.070770   0.001060   0.003646   0.006287
14459  20200531  서울특별시   중구   회현동  ...   0.071026   0.000000   0.009057   0.001663

[14460 rows x 13 columns]

In [12]:
# 차후 업종별 상관계수 및 회귀 모델용 데이터 변수
gs_row = gs.copy()

# GS 데이터 필요 column 줄이기 (구조방정식, 매개효과 용 변수)
gs.columns=['STD_DD', 'HGU_NM', 'CTGG_NM', 'HDONG_NM', 'AMT_IND', 'LCLS_10_P', 'LCLS_20_P',
            'LCLS_30_P', 'LCLS_40_P', 'LCLS_50_P', 'LCLS_60_P', 'LCLS_70_P', 'LCLS_80_P']
gs = gs[['STD_DD', 'HGU_NM', 'HDONG_NM', 'AMT_IND']]
gs

STD_DD HGU_NM HDONG_NM   AMT_IND
0      20190201  대구광역시     고산1동  1.064336
1      20190201  대구광역시     고산2동  1.076418
2      20190201  대구광역시     고산3동  0.856071
3      20190201  대구광역시      두산동  0.724868
4      20190201  대구광역시     만촌1동  1.155728
...         ...    ...      ...       ...
14455  20200531  서울특별시      중림동  1.595038
14456  20200531  서울특별시      청구동  1.389184
14457  20200531  서울특별시       필동  0.714265
14458  20200531  서울특별시      황학동  1.134190
14459  20200531  서울특별시      회현동  0.773963

[14460 rows x 4 columns]

In [13]:
# 유통 GS 품목별 대분류 list dictionary
circulation_list = {'mul_LCLS_10_P':'식사',
                    'mul_LCLS_20_P':'간식',
                    'mul_LCLS_30_P':'마실거리',
                    'mul_LCLS_40_P':'홈&리빙',
                    'mul_LCLS_50_P':'헬스&뷰티',
                    'mul_LCLS_60_P':'취미&여가활동_비중',
                    'mul_LCLS_70_P':'사회활동_비중',
                    'mul_LCLS_80_P':'임신/육아_비중',
                    'mul_LCLS_90_P':'기호품_비중'}

# 유통 GS 품목별 대분류 list dictionary
circulation_list_reverse = {'식사':               'mul_LCLS_10_P',
                            '간식':               'mul_LCLS_20_P',
                            '마실거리' :          'mul_LCLS_30_P',
                            '홈&리빙':            'mul_LCLS_40_P',
                            '헬스&뷰티':          'mul_LCLS_50_P',
                            '취미&여가활동_비중': 'mul_LCLS_60_P',
                            '사회활동_비중':      'mul_LCLS_70_P',
                            '임신/육아_비중':     'mul_LCLS_80_P',
                            '기호품_비중':        'mul_LCLS_90_P'}

## 코로나 데이터 (2020)
- 출처 : https://github.com/jooeungen/coronaboard_kr // CoronaBoard_kr 깃허브 
- 출처 : http://ncov.mohw.go.kr/ // corona board에서 참조한 질병관리본부 링크
- <b>CoronaBoard data - Available Commercial & Non-commercial</b>
  - kr_daily.csv (코로나 일일 확진자 수 - 누적) (원본)
  - kr_regional_data.csv (코로나 지역별  일일 확진자 수 - 누적 )(원본)

- <b>사용데이터 : 코로나 일일 확진자 수 데이터 (전국, 서울, 대구)</b>
  - <b>위 출처 데이터를 1차 가공(차분)한 데이터 (누적 -> 일일) </b>
  - 기간 : 20200201 ~ 20200601 데이터 사용
  - covid_all.csv : 전국 코로나 일일 확진자 수
  - corona_seoul.csv : 서울지역 코로나 일일 확진자 수
  - corona_daegu.csv : 대구지역 코로나 일일 확진자 수
  - 코로나 데이터는 하루씩 당겨서 사용
  

In [14]:
# 코로나 전국 확진자 데이터
covid = pd.read_csv("/content/drive/My Drive/big_contest/corona/covid_all.csv", index_col=0, encoding="cp949")

# 서울, 대구 각각의 지역 확진자 데이터
covid_seoul = pd.read_csv("/content/drive/My Drive/big_contest/corona/corona_seoul.csv", index_col=0, encoding="cp949")
covid_daegu = pd.read_csv("/content/drive/My Drive/big_contest/corona/corona_daegu.csv", index_col=0, encoding="cp949")

# 전국 코로나 데이터
covid = covid.reset_index()
covid.columns = ['STD_DD','confirmed']
covid

STD_DD  confirmed
0    20200201          4
1    20200202          1
2    20200203          3
3    20200204          0
4    20200205          1
..        ...        ...
179  20200729         28
180  20200730         48
181  20200731         18
182  20200801         36
183  20200802         31

[184 rows x 2 columns]

In [15]:
# corona 지역 데이터 (서울, 대구)
covid_seoul = covid_seoul[['date','confirmed_diff']]
covid_seoul.columns=['STD_DD','seoul_confirmed']
covid_daegu = covid_daegu[['date','confirmed_diff']]
covid_daegu.columns=['STD_DD','daegu_confirmed']

# 서울, 대구 코로나 데이터
print(covid_seoul.shape, covid_daegu.shape)
covid_seoul.head(3)

(121, 2) (121, 2)


STD_DD  seoul_confirmed
1  20200201                0
2  20200202                1
3  20200203                0

## 휴일여부 데이터 (2019, 2020)
- 2019년, 2020년 휴일 여부 데이터
- 직접 데이터 수작업 작성
- 명절, 연휴, 대체휴일 등 반영

In [16]:
# 2019, 2020 휴일 여부 데이터
holiday = pd.read_csv("/content/drive/My Drive/big_contest/holiday/holiday.csv")

print(holiday.shape)
holiday

(241, 3)


date  day  weekend
0    20190201    4        0
1    20190202    5        1
2    20190203    6        1
3    20190204    0        1
4    20190205    1        1
..        ...  ...      ...
236  20200527    2        0
237  20200528    3        0
238  20200529    4        0
239  20200530    5        1
240  20200531    6        1

[241 rows x 3 columns]

In [17]:
# 컬럼명 변경, type 변경
holiday.columns = ['STD_DD','day','weekend']

holiday['STD_DD'] = holiday['STD_DD'].astype(str)
holiday['STD_DD'] = holiday['STD_DD'].str.replace("-","")
holiday['STD_DD'] = holiday['STD_DD'].astype(int)
holiday

STD_DD  day  weekend
0    20190201    4        0
1    20190202    5        1
2    20190203    6        1
3    20190204    0        1
4    20190205    1        1
..        ...  ...      ...
236  20200527    2        0
237  20200528    3        0
238  20200529    4        0
239  20200530    5        1
240  20200531    6        1

[241 rows x 3 columns]

## 서울 대구 행정동 리스트
- 서울 중구/노원구 행정동 리스트
- 대구 중구/수성구 행정동 리스트

In [18]:
# 분석의 편의를 위해 서울시 중구, 서울시 노원구, 대구시 중구, 대구시 수성구 동별로 리스트 생성
seoul_jung_list = ['소공동', '회현동', '명동', '필동', '장충동', '광희동', '을지로동', '신당동', '다산동','약수동', '청구동', '신당5동', '동화동', '황학동', '중림동']
seoul_nowon_list = ['월계1동', '월계2동', '월계3동','공릉1동', '공릉2동', '하계1동', '하계2동', '중계본동', '중계1동', '중계4동', '중계2.3동','상계1동', '상계2동', '상계3.4동', '상계5동', '상계6.7동', '상계8동', '상계9동','상계10동']
daegu_jung_list = ['동인동', '삼덕동', '성내1동', '성내2동', '성내3동', '대신동', '남산1동', '남산2동', '남산3동', '남산4동', '대봉1동', '대봉2동']
daegu_suseong_list = ['범어1동', '범어2동', '범어3동', '범어4동', '만촌1동', '만촌2동', '만촌3동', '수성1가동', '수성2.3가동', '수성4가동', '황금1동', 
                      '황금2동', '중동', '상동', '파동', '두산동', '지산1동', '지산2동', '범물1동', '범물2동', '고산1동', '고산2동', '고산3동']

# 4.부트스트래핑
- 비정규성 분포에서 매개효과 가설검증(test)을 하기 위한 Bootstraping 진행
- 매개효과는 2개의 통계량을 곱한 값이기 때문에 중심극한 정리로 정규성을 가정불가
- 따라서 해당 데이터를 가지고 10000번의 30% random sampling을 통해 모집단을 대표할 수 있는 분포를 생성
- 그 후 95% 신뢰구간에서 0이 포함된 매개효과 통계량을 통계적으로 유의하지 않다고 판단

- 매개효과 동리스트 결과 csv
  - mediation_result.csv

- 부트스트래핑 진행 output csv
  - 편의점휴일부트스트래핑_신뢰95_비표준화_10000.csv

In [19]:
# 매개효과 동 리스트 결과
mediation_result = pd.read_csv('/content/drive/My Drive/big_contest/bootstraping/mediation_result.csv', encoding='cp949')

# 매개 효과 존재 리스트
card_weekday_list = mediation_result.iloc[0, ].values[0].split(',')
card_weekend_list = mediation_result.iloc[1, ].values[0].split(',')
gs_weekday_list = mediation_result.iloc[2, ].values[0].split(',')
gs_weekend_list = mediation_result.iloc[3, ].values[0].split(',')

print('카드 평일 매개 존재 검증 리스트 : ', len(card_weekday_list), card_weekday_list,'\n')
print('카드 휴일 매개 존재 검증 리스트 : ', len(card_weekend_list), card_weekend_list,'\n')
print('편의점 평일 매개 존재 검증 리스트 : ', len(gs_weekday_list), gs_weekday_list, '\n')
print('편의점 휴일 매개 존재 검증 리스트 : ', len(gs_weekend_list), gs_weekend_list, '\n')

카드 평일 매개 존재 검증 리스트 :  69 ['고산1동', '고산2동', '고산3동', '두산동', '만촌1동', '만촌2동', '만촌3동', '범물1동', '범물2동', '범어1동', '범어2동', '범어3동', '범어4동', '상동', '수성1가동', '수성2.3가동', '수성4가동', '중동', '지산1동', '지산2동', '파동', '황금1동', '황금2동', '남산1동', '남산2동', '남산3동', '남산4동', '대봉1동', '대봉2동', '대신동', '동인동', '삼덕동', '성내1동', '성내2동', '성내3동', '공릉1동', '공릉2동', '상계10동', '상계1동', '상계2동', '상계3.4동', '상계5동', '상계6.7동', '상계8동', '상계9동', '월계1동', '월계2동', '월계3동', '중계1동', '중계2.3동', '중계4동', '중계본동', '하계1동', '하계2동', '광희동', '다산동', '동화동', '명동', '소공동', '신당5동', '신당동', '약수동', '을지로동', '장충동', '중림동', '청구동', '필동', '황학동', '회현동'] 

카드 휴일 매개 존재 검증 리스트 :  69 ['고산1동', '고산2동', '고산3동', '두산동', '만촌1동', '만촌2동', '만촌3동', '범물1동', '범물2동', '범어1동', '범어2동', '범어3동', '범어4동', '상동', '수성1가동', '수성2.3가동', '수성4가동', '중동', '지산1동', '지산2동', '파동', '황금1동', '황금2동', '남산1동', '남산2동', '남산3동', '남산4동', '대봉1동', '대봉2동', '대신동', '동인동', '삼덕동', '성내1동', '성내2동', '성내3동', '공릉1동', '공릉2동', '상계10동', '상계1동', '상계2동', '상계3.4동', '상계5동', '상계6.7동', '상계8동', '상계9동', '월계1동', '월계2동', '월계3동', '중계1동', '중계2.3동', '중계4동

## 편의점데이터부트스트래핑-휴일
- gs_weekend_list : 편의점 휴일


In [ ]:
# 편의점 + 코로나 + 휴일 + 유동인구
all=pd.merge(covid, gs, on='STD_DD')
all=pd.merge(all, skt_2020, on=['STD_DD','HDONG_NM'])
all=pd.merge(all, holiday, on='STD_DD')
all=all[['STD_DD', 'confirmed','AMT_IND', 'sum_20', 'HGU_NM', 'HDONG_NM','day','weekend']]
all

STD_DD  confirmed   AMT_IND     sum_20 HGU_NM HDONG_NM  day  weekend
0     20200201          4  0.854239  132881.50  대구광역시     고산1동    5        1
1     20200201          4  1.080543  232618.75  대구광역시     고산2동    5        1
2     20200201          4  0.759088   49579.14  대구광역시     고산3동    5        1
3     20200201          4  0.911892   96455.73  대구광역시      두산동    5        1
4     20200201          4  1.124104  104249.83  대구광역시     만촌1동    5        1
...        ...        ...       ...        ...    ...      ...  ...      ...
7255  20200531         39  1.595038   21982.16  서울특별시      중림동    6        1
7256  20200531         39  1.389184   15985.86  서울특별시      청구동    6        1
7257  20200531         39  0.714265   53756.05  서울특별시       필동    6        1
7258  20200531         39  1.134190   28761.22  서울특별시      황학동    6        1
7259  20200531         39  0.773963  104740.08  서울특별시      회현동    6        1

[7260 rows x 8 columns]

In [ ]:
# 편의점 휴일 부트스트래핑

import numpy as np
import scipy.stats
from tqdm.notebook import tqdm

def mean_confidence_interval(data, confidence=0.95):
    a = 1.0 * np.array(data)
    n = len(a)
    m, se = np.mean(a), scipy.stats.sem(a)
    h = se * scipy.stats.t.ppf((1 + confidence) / 2., n-1)
    return m, m-h, m+h

# 부트스트래핑 횟수
iterationNum = 10000

# 부트스트래핑 휴일 편의점

data=[]
data=pd.DataFrame(data)
for i in tqdm(gs_weekend_list):
    coef=[]
    coef=pd.DataFrame(coef)
    a=all[(all.HDONG_NM==i) & (all.weekend==1)][['confirmed','AMT_IND','sum_20']] #휴일은 all.weekend==1
    for n in tqdm(range(iterationNum)):
        b=a.sample(round(len(a)*0.3)) 

        # 확진자 - 유동인구
        model = smf.ols(formula = 'sum_20 ~ confirmed', data = b)
        result = model.fit()
        df = pd.read_html(result.summary().tables[1].as_html(),header=0,index_col=0)[0]
        coef.loc[n,i+'confirmed'+"_"+'sum_20'+"_coef"]=df.loc['confirmed','coef']

        #유동인구 AMT_IND(확진자 통제)
        model = smf.ols(formula = 'AMT_IND ~ confirmed +sum_20', data = b)
        result = model.fit()
        df = pd.read_html(result.summary().tables[1].as_html(),header=0,index_col=0)[0]
        coef.loc[n,i+"_sum_20_AMT_IND_partial_coef"]=df.loc['sum_20','coef']

    coef['ac']=coef[i+"confirmed_sum_20_coef"]*coef[i+'_sum_20_AMT_IND_partial_coef']
    coef=coef['ac']
    c=mean_confidence_interval(coef,confidence=0.95)[1]
    d=mean_confidence_interval(coef,confidence=0.95)[2]
    print(np.mean(coef))
    print(i,c,d)
    data.loc[i, '하한']=c
    data.loc[i, '상한']=d
    data.loc[i, 'mean']=coef.mean()
    
data = pd.DataFrame(data)
data['HDONG_NM']=gs_weekend_list
# data.to_csv('편의점휴일부트스트래핑_신뢰95_비표준화.csv', index=False, encoding='cp949')
data.to_csv('/content/drive/My Drive/big_contest/bootstraping/편의점휴일부트스트래핑_신뢰95_비표준화_10000.csv', index=False, encoding='cp949')
print('저장완료')


-0.0002114480301798565
고산1동 -0.00021521102817049579 -0.00020768503218921724



-0.00018758981942730148
고산2동 -0.00019334704919168113 -0.00018183258966292183



-0.00011689739453437684
고산3동 -0.00011940078399424745 -0.00011439400507450622



-0.001679901006830452
두산동 -0.0017091923803470117 -0.0016506096333138921



-0.0004131052062672035
만촌1동 -0.0004223893062029067 -0.0004038211063315003



-0.000679820698303504
만촌3동 -0.0006938680946236416 -0.0006657733019833664



-0.00033422649917360693
범물1동 -0.0003419177969936712 -0.00032653520135354265



-8.266297838353154e-05
범물2동 -8.835733218700334e-05 -7.696862458005975e-05



-0.0009433906899789193
범어1동 -0.0009603840974935012 -0.0009263972824643373



-0.0007317195547063324
범어2동 -0.0007436064180870297 -0.0007198326913256352



-0.000827265765965938
범어3동 -0.0008411947712131831 -0.0008133367607186929



-0.00112249722308947
범어4동 -0.0011389670362429729 -0.001106027409935967



-0.00033767343574338693
수성1가동 -0.0003435105869638009 -0.00033183628452297296



-0.000434568510224019
수성2.3가동 -0.0004412930819419001 -0.0004278439385061379



-0.0004138046006883393
수성4가동 -0.0004210003420797619 -0.00040660885929691664



-0.0003440730486458939
중동 -0.00034965178141511326 -0.0003384943158766746



-0.00013849371812223461
지산1동 -0.00014410004315240616 -0.00013288739309206307



-5.476145186099782e-06
지산2동 -8.451867045646293e-06 -2.5004233265532717e-06



-0.0003869778022169491
황금1동 -0.0003955415705065254 -0.00037841403392737283



-0.0002665016231151242
황금2동 -0.0002733743988724825 -0.0002596288473577659



-0.0010006270187092
남산1동 -0.0010181602916473128 -0.0009830937457710872



-0.0003520687237173242
남산4동 -0.0003589791857672739 -0.00034515826166737454



-0.00026003748614217146
동인동 -0.0002640845728316301 -0.0002559903994527128



-0.0006504711899663762
삼덕동 -0.000661566478295357 -0.0006393759016373954



-0.001921255572842821
성내1동 -0.0019559967857208952 -0.001886514359964747



-0.001044863029197747
성내2동 -0.0010638489881596193 -0.0010258770702358745



-0.0007449146807831911
성내3동 -0.0007598366763894533 -0.0007299926851769288



-0.00017048104057602057
공릉1동 -0.0001736351727394731 -0.00016732690841256804



-6.658535657133779e-05
공릉2동 -6.938850940120612e-05 -6.378220374146946e-05



-0.00027918652585529
상계10동 -0.0002841885869563937 -0.00027418446475418627



-0.00021455046888436445
상계1동 -0.00021848049594182302 -0.00021062044182690588



-0.0004857513791751487
상계2동 -0.0004943994171933526 -0.0004771033411569448



2.0905649610067345e-05
상계3.4동 1.9806844414825145e-05 2.2004454805309544e-05



-9.633538518040612e-05
상계5동 -9.846712004326808e-05 -9.420365031754415e-05



-0.0003003689478201877
상계6.7동 -0.00030599733436510545 -0.00029474056127526995



-0.0002631084707160734
상계9동 -0.00026803985997403613 -0.00025817708145811066



-0.0001093444551446287
월계1동 -0.00011215488635401698 -0.00010653402393524041



-7.3802067333948205e-06
월계2동 -8.679566101681241e-06 -6.080847365108401e-06



-8.454050561773018e-05
월계3동 -8.737940947297142e-05 -8.170160176248894e-05



-0.00019091214422221444
중계1동 -0.0001946199007082411 -0.00018720438773618778



-0.00026188911765387147
중계2.3동 -0.00026896427966496746 -0.00025481395564277547



-0.00029778416601824504
중계4동 -0.0003035268359571605 -0.00029204149607932957



-8.819548240894068e-05
중계본동 -9.212146622901113e-05 -8.426949858887023e-05



-0.0004348550712961296
하계1동 -0.00044224599834702914 -0.00042746414424523003



-0.0002009565682588657
하계2동 -0.00020658317568422589 -0.00019532996083350553



-0.00042863096116567396
광희동 -0.0004390087605776415 -0.0004182531617537064



-0.00013818556268389926
다산동 -0.0001486726576622202 -0.00012769846770557832



-1.951752975234151e-05
동화동 -2.225641905172682e-05 -1.6778640452956203e-05



-0.0005930346448112019
명동 -0.0006039163770799468 -0.000582152912542457



-0.00046088879809569707
소공동 -0.0004707350590092462 -0.00045104253718214796



-0.0001287368257295036
신당5동 -0.0001324243951286887 -0.0001250492563303185



-0.0007363888076921499
신당동 -0.0007557886241452801 -0.0007169889912390196



-3.620371231433037e-05
약수동 -3.906607830009318e-05 -3.334134632856756e-05



-0.000780179332717208
을지로동 -0.0007982236827496728 -0.0007621349826847433



-0.00010493948663627473
장충동 -0.00010758629184558083 -0.00010229268142696862



-0.00027994706994882056
중림동 -0.00028637554086186724 -0.0002735185990357739



-0.0001517962926423249
청구동 -0.00015596254487034786 -0.00014763004041430192



-0.00029504649997063687
필동 -0.0003016482744482366 -0.0002884447254930371



-0.00020858590823777078
황학동 -0.00021343043030302931 -0.00020374138617251225



-0.00014077594900575827
회현동 -0.00015139041944418292 -0.00013016147856733363

저장완료


In [20]:
# 편의점 부트스트래핑 기각 결과
gs_boot_weekend = pd.read_csv('/content/drive/My Drive/big_contest/bootstraping/편의점휴일부트스트래핑_신뢰95_비표준화_10000.csv', encoding='cp949')
gs_boot_weekend['ac'] = gs_boot_weekend['하한']*gs_boot_weekend['상한'] # 사이에 0이 있는거 찾기 위함 곱해서 음수면 사이에 0있음

print('휴일 0 포함', gs_boot_weekend[gs_boot_weekend['ac']<=0].HDONG_NM)

휴일 0 포함 Series([], Name: HDONG_NM, dtype: object)
